File: src/0051-N-Queens/0051.cpp
#include <iostream>
#include <vector>
using namespace std;


static int x = [](){std::ios::sync_with_stdio(false);cin.tie(0);return 0;}();
class Solution 
{
public:
    vector<vector<string>> solveNQueens(int n) 
    {
        col = vector<bool>(n, false);
        dia1 = vector<bool>(2*n - 1, false);
        dia2 = vector<bool>(2*n - 1, false);
        vector<int> row;
        putQueen(n, 0, row);
        return result;
    }
private:
    vector<vector<string>> result;
    vector<bool> col, dia1, dia2;

    void putQueen(int n, int index, vector<int>& row)
    {
        if (index == n)
        {
            result.push_back(generateBoard(n, row));
            return;
        }

        for (unsigned int i = 0; i < n; ++i)
        {
            if (!col[i] and !dia1[index + i] and !dia2[index - i + n - 1])
            {
                row.push_back(i);
                col[i] = true;  dia1[index + i] = true;  dia2[index - i + n - 1] = true;
                putQueen(n, index + 1, row);
                col[i] = false; dia1[index + i] = false; dia2[index - i + n - 1] = false;
                row.pop_back();
            }
        }
    }

    vector<string> generateBoard(int n, vector<int>& row)
    {
        vector<string> board(n, string(n, '.'));
        for (unsigned int i = 0; i < n; ++i)
        {
            board[i][row[i]] = 'Q';
        }
        return board;
    }
};

int main()
{
    int n = 4;
    vector<vector<string>> result = Solution().solveNQueens(n);
    for (unsigned int i = 0; i < result.size(); ++i)
    {
        for (unsigned int j = 0; j < n; ++j)
        {
            cout << result[i][j] << endl;
        }
        cout << endl;
    }
}

File: src/0051-N-Queens/0051.py
class Solution:
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        result, R, col, dia1, dia2 = list(), list(), list(), list(), list()
        col = [0 for i in range(n)]
        dia1 = [0 for i in range(2*n - 1)]
        dia2 = [0 for i in range(2*n - 1)]

        def generateBoard(m, row):
            board = [str() for i in range(n)]           
            for i in range(m):
                board[i] = row[i]*'.' + 'Q' + '.'*(m - row[i] - 1)

            return board

        def putQueen(m, index, row):
            if index == m:
                result.append(generateBoard(m, row))
                return

            for i in range(m):
                if not (col[i] or dia1[index + i] or dia2[index - i + m - 1]):
                    row.append(i)
                    col[i], dia1[index + i], dia2[index - i + m - 1] = 1, 1, 1
                    putQueen(m, index + 1, row)
                    col[i], dia1[index + i], dia2[index - i + m - 1] = 0, 0, 0
                    row.pop()

        putQueen(n, 0, R)
        return result

if __name__ == '__main__':
    n = 4
    print(Solution().solveNQueens(n))